p4-16/spec/P4-16-spec.mdk

```diff
@@ -6882,11 +6920,219 @@ control ingress(inout Header_t h, inout Meta_t m,
+the `entries` table property, then the compiler issues a warning if
+any two entries for the same table have equal priority values.  Both
+P4Runtime and TDI leave it unspecified which entry is the winner if a
+lookup key matches multiple keys that all have the same priority,
+hence a warning is useful to less experienced developers that are
+unfamiliar with this unspecified behavior.
+
+If the annotation `@noWarn("duplicate_priorities")` is used on the
+`entries` table property, then no warnings of this type are ever
+issued by the compiler.  Using equal priority values for multiple
+entries in the same table is sometimes useful in reducing the number
+of hardware updates required when adding entries to such a table.
+
+If the annotation `@noWarn("entries_out_of_priority_order")` is _not_
+used on the `entries` table property, then the compiler issues a
+warning if:
+
+- If `largest_priority_wins` is `true` for the table, and there is any
+  pair of consecutive entries where `prio[j] < prio[j+1]`, then a
+  warning is issued for that pair of entries.
+- If `largest_priority_wins` is `false` for the table, and there is any
+  pair of consecutive entries where `prio[j] > prio[j+1]`, then a
+  warning is issued for that pair of entries.
+
+This warning is useful to developers that want the order that entries
+appear in the source code to match the relative priority of entries in
+the target device.
+
+If the annotation `@noWarn("entries_out_of_priority_order")` is used
+on the `entries` table property, then no warnings of this type are
+ever issued by the compiler for this table.  This option is provided
+for developers who explicitly choose to specify entries in an order
+that does not match their relative priority order.
+
+The following example is the same as the first example in section
+[#sec-entries], except for the definition of table `t_exact_ternary`
+shown below.
+
+~ Begin P4Example
+table t_exact_ternary {
+    key = {
+        h.h.e : exact;
+        h.h.t : ternary;
+    }
+
+    actions = {
+        a;
+        a_params;
+    }
+
+    default_action = a;
+
+    largest_priority_wins = false;
+    priority_delta = 10;
+    @noWarn("duplicate_priorities")
+    entries = {
+        const priority=10 (0x01, 0x1111 &&& 0xF   ) : a_params(1);
+                          (0x02, 0x1181           ) : a_params(2); // priority=20
+                          (0x03, 0x1000 &&& 0xF000) : a_params(3); // priority=30
+        const             (0x04, 0x0210 &&& 0x02F0) : a_params(4); // priority=40
+              priority=40 (0x04, 0x0010 &&& 0x02F0) : a_params(5);
+                          (0x06, _                ) : a_params(6); // priority=50
+    }
+}
+~ End P4Example
+
+The entries that do not have an explicit priority specified will be
+assigned the priority values shown in the comments, because
+`priority_delta` is 10, and because of those entries that do have
+priority values specified.
```

Comment:
these priorities do not change when the table is modified.

Reply:
I added a short paragraph to the text explaining this.  Rather than saying that the priorities do not change when the table is modified, I left it up to the behavior defined by the control plane API in use.  This leaves open the possibility for a control plane API to define mechanisms to modify the priority values if entries, if this is deemed useful, without contradicting the language spec.